framework/calico_framework.py

```diff
@@ -0,0 +1,707 @@
+        Useful for explaining a test-scope failure (vs. a task failure).
+        """
+
+        self.killed = False
+        """
+        Flag, when set to True, will indicate that this test is being shutdown.
+        """
+
+    def __repr__(self):
+        s_repr = "Test(%s" % self.name
+        if self.timeout is not None:
+            s_repr += (", timeout=%d" % (time.time() - self.timeout))
+        s_repr += ")"
+
+        return s_repr
+
+    @property
+    def complete(self):
+        """
+        True if the test is no longer running. This means killed tests too
+        """
+        return self.killed or self.tasks_with_state(
+            UNFINISHED_TASK_STATES) == []
+
+    def can_run_on(self, offers):
+        """
+        Returns true if the test can run on the offers passed in.
+        Offers passed in will be ignored if they are already reserved.
+        This function also reserves offers and plans tasks as it goes along.
+        It will rollback reservations if it realizes it can't finish
+        planning all tasks.
+        """
+        # First sort the tasks by slave
+        tasks_by_slave = {}
+        for task in [task for task in self.tasks if task.slave is not None]:
+            tasks_by_slave.setdefault(task.slave, []).append(task)
+
+        # Sort those groups by most tasks to least tasks
+        tasks_grouped_by_slave = sorted(tasks_by_slave.values(),
+                                        lambda x, y: len(y) - len(x))
+
+        # Quit early if there isn't enough unique slaves
+        if len(tasks_grouped_by_slave) > len(offers):
+            return False
+
+        # Then find the remaining tasks without a specified slave requirement
+        anywhere_tasks = [task for task in self.tasks if task.slave is None]
+
+        # Sort the offers by resources (this assumes each offer is from a
+        # unique slave
+        offers_by_tasks_launchable = sorted(offers, lambda x, y:
+        cmp(y.num_tasks_launchable, x.num_tasks_launchable))
+
+        # We now have the offers sorted by largest to smallest, and the
+        # tasks-per-slave grouped from largest to smallest.
+        try:
+            for offer, tasks in zip(offers_by_tasks_launchable,
+                                    tasks_by_slave.values()):
+                if offer.num_tasks_launchable < len(tasks):
+                    raise NotEnoughResources(
+                        "Need a larger offer to meet slave-id specifications")
+
+                # We can fit the tasks for this slave on this offer.
+                for task in tasks:
+                    _log.debug("\t\tPlanning %s for %s", task, offer)
+                    task.slave_id = offer.offer.slave_id.value
+                    offer.reserved_by = self
+                    offer.tasks.append(task)
+
+            # Squeeze the anywhere offers in
+            for anywhere_task in anywhere_tasks:
+                for offer in offers:
```

Comment:
This is bugged.  Even after you've found a successful offer, you'll continue to examine offers.

A better structure would be to use `anywhere_tasks` as a stack.  As you're traversing the offers, `pop()` tasks off anywhere_tasks until the offer is full, or you run out of tasks.  You can even do this in the `zip`ped loop above.

You also end with a much cheaper operation to tell whether you assigned them all:  `if anywhere_tasks:`